source/crud/crud.rst

```diff
@@ -415,8 +415,10 @@ Read
      * can write temporary data to disk while executing the find operation.
      *
      * This option is sent only if the caller explicitly provides a value. The default
-     * is to not send a value. For servers < 3.2, this option is ignored and not sent
-     * as allowDiskUse does not exist in the OP_QUERY wire protocol.
+     * is to not send a value.
+     *
+     * This option is only supported by servers >= 4.4. Older servers >= 3.2 will report an error for using this option.
+     * For servers < 3.2, the driver MUST raise an error if the caller explicitly provides a value.
```

Comment:
Reading [SPEC-1673](https://jira.mongodb.org/browse/SPEC-1673), I think you may have missed the context for the current behavior from the original code review for [SPEC-1418](https://jira.mongodb.org/browse/SPEC-1418). See: https://github.com/mongodb/specifications/pull/692#discussion_r366563400 and follow-up PR in #716.

To recap: The behavior to silently ignore `allowDiskUse` for older server versions using OP_QUERY is more akin to `maxAwaitTimeMS`, which we silently ignore for OP_GET_MORE. This is contrast to `hint`, where we want to guarantee users see an error (either from client or server) because the option could entirely change the results returned by the server based on the index used.

In the case of `allowDiskUse`, failure to apply this option to old queries at worse could lead to a memory error. With `maxAwaitTimeMS`, a query without it would theoretically return the same result if the cursor was iterated further.

cc: @jyemin in case he has additional context to share

Reply:
I did miss the context of that PR, though having read through it I can't say I agree with its conclusion. Specifically, the rationale hinges on the assertion that "Application sets AllowDiskUse and server does not support it and ignores it: not a problem", which I do not believe to be true from several angles:
- a user specifying `allowDiskUse` against a server that doesn't support it is _always_ a bug in the user's code. We should ideally notify them of this ASAP.
- It's inconsistent from an API perspective for the driver to not report errors conditionally when an option isn't supported.
- "you're going to get the exact same result set whether you specify AllowDiskUse or not (assuming sufficient memory)." - the "assuming" clause assumes too much here, seeing as the purpose of the option is to obviate concerns about having sufficient memory. Without the assumption, the presence of the option absolutely can affect the outcome of a query.
- It would be surprising behavior for a user to have ticked the option for paging to disk to see an out of memory error. It would not be surprising behavior to see that they got an error for specifying an invalid option.

I do agree that given 3.0 has been EOL for a while, we shouldn't necessarily be making lots of exceptions for supporting it. The fact of the matter is that we already are doing that in other places though (e.g. `hint`), so we should be consistent about it.

Reply:
No objections on my end. If this is merged, I'd suggest requesting language tickets ASAP so some drivers have a chance to make the corresponding change in their 4.4-compat release (alongside previous work for SPEC-1418).

I stand by my suggestion to split up client and server error tests below.

Reply:
Sounds good. I think the others involved in the original decision will probably want to weigh in here before we can merge anything though.

Reply:
I'm fine with the change as it is, let's see if @jyemin has any opinion here

Reply:
There are enough cooks in this kitchen.  I'll leave it to you all to make the call.